File: app/src/main/java/co/com/sersoluciones/facedetectorser/fragments/SaveImageFragment.java
package co.com.sersoluciones.facedetectorser.fragments;

import android.animation.Animator;
import android.animation.AnimatorListenerAdapter;
import android.annotation.SuppressLint;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.Matrix;
import android.graphics.drawable.BitmapDrawable;
import android.net.Uri;
import android.os.AsyncTask;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.design.widget.FloatingActionButton;
import android.support.v4.app.Fragment;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.ImageView;

import java.io.ByteArrayOutputStream;
import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;

import co.com.sersoluciones.facedetectorser.FaceTrackerActivity;
import co.com.sersoluciones.facedetectorser.R;

/**
 * Created by Ser Soluciones SAS on 28/12/2017.
 * www.sersoluciones.com - contacto@sersoluciones.com
 **/
public class SaveImageFragment extends Fragment {

    Button previousButton;
    Button submitButton;

    private static final String ARG_PHOTO_PATH = "mPhotoPath";
    private static final String ARG_URI = "uri_image";
    ImageView imagePreview;
    private String mPhotoPath;
    private Uri uriImage;
    private Bitmap mBitmap;
    private int rotateImage;
    private View mProgressView;

    public SaveImageFragment() {
    }

    /**
     * Use this factory method to create a new instance of
     * this fragment using the provided parameters.
     *
     * @param mPhotoPath Parameter 1.
     * @return A new instance of fragment ValidityFragment.
     */
    // TODO: Rename and change types and number of parameters
    public static SaveImageFragment newInstance(String mPhotoPath) {
        SaveImageFragment fragment = new SaveImageFragment();
        Bundle args = new Bundle();
        args.putString(ARG_PHOTO_PATH, mPhotoPath);
        fragment.setArguments(args);
        return fragment;
    }

    public static SaveImageFragment newInstance(Uri uri) {
        SaveImageFragment fragment = new SaveImageFragment();
        Bundle args = new Bundle();
        args.putString(ARG_URI, uri.toString());
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        mPhotoPath = "";
        rotateImage = 0;
        mBitmap = null;
        if (getArguments() != null) {
            if (getArguments().containsKey(ARG_PHOTO_PATH))
                mPhotoPath = getArguments().getString(ARG_PHOTO_PATH);
            else
                uriImage = Uri.parse(getArguments().getString(ARG_URI));
        }
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        // Inflate the layout for this fragment
        View view = inflater.inflate(R.layout.fragment_save_photo, container, false);

        imagePreview = view.findViewById(R.id.image_preview);
        submitButton = view.findViewById(R.id.submit_button);
        submitButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                saveImage();
            }
        });
        previousButton = view.findViewById(R.id.previous_button);
        mProgressView = view.findViewById(R.id.progress);

        FloatingActionButton rotateButton = view.findViewById(R.id.fab_rotate);
        rotateButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                rotateLogo();
            }
        });
        previousButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                removeFragment();
            }
        });

        Bitmap bitmap;
        if (!mPhotoPath.isEmpty()) {
            bitmap = BitmapFactory.decodeFile(mPhotoPath);
            imagePreview.setImageBitmap(bitmap);
        } else {
            imagePreview.setImageURI(uriImage);
            if (mPhotoPath.isEmpty())
                mPhotoPath = uriImage.getPath();
            //bitmap = BitmapFactory.decodeFile(mPhotoPath);
            //imagePreview.setImageBitmap(bitmap);
        }
        return view;
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
    }

    public void saveImage() {
        showProgress(true);
        if ((getActivity()) != null) {
            if (rotateImage > 0) {
               new SaveImageAsyncTask().execute();
            }else{
                ((FaceTrackerActivity) getActivity()).returnURIImage(mPhotoPath);
            }
        }
    }

    public void removeFragment() {
        if (getActivity() != null) {
            ((FaceTrackerActivity) getActivity()).removeFragment();
        }
    }

    public void rotateLogo() {
        showProgress(true);
        rotateImage += 90;
        if (rotateImage == 360) rotateImage = 0;
        mBitmap = ((BitmapDrawable) imagePreview.getDrawable()).getBitmap();
        new RotateAsyncTask().execute(90.0f);
    }

    /**
     * Shows the progress UI and hides the login form.
     */
    private void showProgress(final boolean show) {
        // On Honeycomb MR2 we have the ViewPropertyAnimator APIs, which allow
        // for very easy animations. If available, use these APIs to fade-in
        // the progress spinner.
        int shortAnimTime = getResources().getInteger(android.R.integer.config_shortAnimTime);

        mProgressView.setVisibility(show ? View.VISIBLE : View.GONE);
        mProgressView.animate().setDuration(shortAnimTime).alpha(
                show ? 1 : 0).setListener(new AnimatorListenerAdapter() {
            @Override
            public void onAnimationEnd(Animator animation) {
                mProgressView.setVisibility(show ? View.VISIBLE : View.GONE);
            }
        });
    }

    @SuppressLint("StaticFieldLeak")
    class SaveImageAsyncTask extends AsyncTask<Void, Void, Void> {

        @Override
        protected Void doInBackground(Void... params) {
            ByteArrayOutputStream ostream = new ByteArrayOutputStream();
            mBitmap.compress(Bitmap.CompressFormat.JPEG, 100, ostream);
            try {
                File imageFile = new File(mPhotoPath);
                FileOutputStream fout = new FileOutputStream(imageFile);
                fout.write(ostream.toByteArray());
                fout.close();
            } catch (IOException e) {
                e.printStackTrace();
            }
            return null;
        }

        @Override
        protected void onPostExecute(Void aVoid) {
            super.onPostExecute(aVoid);
            ((FaceTrackerActivity) getActivity()).returnURIImage(mPhotoPath);
        }
    }

    @SuppressLint("StaticFieldLeak")
    class RotateAsyncTask extends AsyncTask<Float, Void, Bitmap> {

        @Override
        protected Bitmap doInBackground(Float... params) {
            return rotateImage(params[0]);
        }

        @Override
        protected void onPostExecute(Bitmap bitmap) {
            super.onPostExecute(bitmap);

            mBitmap = bitmap;
            imagePreview.setImageBitmap(mBitmap);
            showProgress(false);
        }

        Bitmap rotateImage(float angle) {

            Bitmap bitmap = null;
            Matrix matrix = new Matrix();
            matrix.postRotate(angle);
            try {
                bitmap = Bitmap.createBitmap(mBitmap, 0, 0, mBitmap.getWidth(), mBitmap.getHeight(),
                        matrix, false);
            } catch (OutOfMemoryError err) {
                err.printStackTrace();
            }
            return bitmap;
        }
    }
}

File: app/src/main/java/co/com/sersoluciones/facedetectorser/fragments/SaveImageFragment.kt
package co.com.sersoluciones.facedetectorser.fragments

import android.animation.Animator
import android.animation.AnimatorListenerAdapter
import android.annotation.SuppressLint
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.graphics.Matrix
import android.graphics.drawable.BitmapDrawable
import android.net.Uri
import android.os.AsyncTask
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import co.com.sersoluciones.facedetectorser.FaceTrackerActivity
import co.com.sersoluciones.facedetectorser.databinding.FragmentSavePhotoBinding
import co.com.sersoluciones.facedetectorser.utilities.DebugLog.logW
import java.io.FileDescriptor

/**
 * Created by Ser Soluciones SAS on 28/12/2017.
 * www.sersoluciones.com - contacto@sersoluciones.com
 */
class SaveImageFragment : Fragment() {

    private var uriImage: Uri? = null
    private var mBitmap: Bitmap? = null
    private var rotateImage = 0

    private lateinit var binding: FragmentSavePhotoBinding

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        rotateImage = 0
        mBitmap = null
        arguments?.let {
            if (it.containsKey(ARG_URI)) uriImage = Uri.parse(arguments!!.getString(ARG_URI))
            logW("uriImage: $uriImage")
        }

    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        binding = FragmentSavePhotoBinding.inflate(inflater)
        return binding.root
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)

        binding.submitButton.setOnClickListener { saveImage() }
        binding.fabRotate.setOnClickListener { rotateLogo() }

        binding.previousButton.setOnClickListener { removeFragment() }

        //BitmapFactory.Options options = new BitmapFactory.Options();
        //options.inSampleSize = 2;
//            bitmap = BitmapFactory.decodeFile(uriImage!!.path)

//            val openStream = URL(uriImage!!.toString()).openStream()
//            bitmap = BitmapFactory.decodeStream(openStream)
//
//            val ostream = ByteArrayOutputStream()
//            bitmap.compress(Bitmap.CompressFormat.JPEG, 50, ostream)
//            val decoded = BitmapFactory.decodeStream(ByteArrayInputStream(ostream.toByteArray()))
//            binding.imagePreview.setImageBitmap(decoded)

        val readOnlyMode = "r"
        val image = activity?.contentResolver?.openFileDescriptor(uriImage!!, readOnlyMode).use { pfd ->
            pfd?.let {
                val fileDescriptor: FileDescriptor = it.fileDescriptor
                BitmapFactory.decodeFileDescriptor(fileDescriptor)
            }
        }
        binding.imagePreview.setImageBitmap(image)

    }

    private fun saveImage() {
        showProgress(true)
        if (activity != null) {
            if (rotateImage > 0) {
                SaveImageAsyncTask().execute()
            } else {
                (activity as FaceTrackerActivity?)!!.returnURIImage(uriImage!!)
            }
        }
    }

    private fun removeFragment() {
        if (activity != null) {
            (activity as FaceTrackerActivity?)!!.removeFragment()
        }
    }

    private fun rotateLogo() {
        showProgress(true)
        rotateImage += 90
        if (rotateImage == 360) rotateImage = 0
        mBitmap = (binding.imagePreview.drawable as BitmapDrawable).bitmap
        RotateAsyncTask().execute(90.0f)
    }

    /**
     * Shows the progress UI and hides the login form.
     */
    private fun showProgress(show: Boolean) {
        // On Honeycomb MR2 we have the ViewPropertyAnimator APIs, which allow
        // for very easy animations. If available, use these APIs to fade-in
        // the progress spinner.
        val shortAnimTime = resources.getInteger(android.R.integer.config_shortAnimTime)
        binding.progress.visibility = if (show) View.VISIBLE else View.GONE
        binding.progress.animate().setDuration(shortAnimTime.toLong()).alpha(
                if (show) 1f else 0f).setListener(object : AnimatorListenerAdapter() {
            override fun onAnimationEnd(animation: Animator) {
                binding.progress.visibility = if (show) View.VISIBLE else View.GONE
            }
        })
    }

    @SuppressLint("StaticFieldLeak")
    inner class SaveImageAsyncTask : AsyncTask<Void?, Void?, Void?>() {
        override fun doInBackground(vararg params: Void?): Void? {

            try {
                activity?.contentResolver!!.openOutputStream(uriImage!!).let { out ->
                    mBitmap!!.compress(Bitmap.CompressFormat.JPEG, 100, out)
                }
            } catch (e: Exception) {
                e.printStackTrace()
                logW("Failed to insert image")
            }

            return null
        }

        override fun onPostExecute(aVoid: Void?) {
            super.onPostExecute(aVoid)
            (activity as FaceTrackerActivity?)!!.returnURIImage(uriImage!!)
        }
    }

    @SuppressLint("StaticFieldLeak")
    inner class RotateAsyncTask : AsyncTask<Float?, Void?, Bitmap?>() {

        override fun doInBackground(vararg params: Float?): Bitmap? {
            return rotateImage(params[0]!!)
        }

        override fun onPostExecute(bitmap: Bitmap?) {
            super.onPostExecute(bitmap)
            mBitmap = bitmap
            binding.imagePreview.setImageBitmap(mBitmap)
            showProgress(false)
        }

        private fun rotateImage(angle: Float): Bitmap? {
            var bitmap: Bitmap? = null
            val matrix = Matrix()
            matrix.postRotate(angle)
            try {
                bitmap = Bitmap.createBitmap(mBitmap!!, 0, 0, mBitmap!!.width, mBitmap!!.height,
                        matrix, false)
            } catch (err: OutOfMemoryError) {
                err.printStackTrace()
            }
            return bitmap
        }


    }

    companion object {
        private const val ARG_URI = "uri_image"

        /**
         * Use this factory method to create a new instance of
         * this fragment using the provided parameters.
         *
         * @param uri uri from image
         * @return A new instance of fragment ValidityFragment.
         */
        fun newInstance(uri: Uri): SaveImageFragment {
            val fragment = SaveImageFragment()
            val args = Bundle()
            args.putString(ARG_URI, uri.toString())
            fragment.arguments = args
            return fragment
        }
    }
}